src/interp.rs

```diff
@@ -34,36 +34,89 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
+use byteorder::{ByteOrder, LittleEndian};
 use std::process::exit;
 use rustc::ty;
 use rustc::ty::{Ty, TyCtxt, TypeVariants};
 use rustc::hir::def_id::DefId;
+use rustc::middle::const_val::ConstVal;
 use rustc::mir::{
     START_BLOCK, Mir, BasicBlock, BasicBlockData, Statement, StatementKind,
     Rvalue, Place, Terminator, TerminatorKind, AggregateKind, Operand, Constant,
-    Literal
+    Literal, BinOp
 };
+use syntax::ast::IntTy;
 
 use machine;
 use machine::{Machine, Address};
-use primitive::ByteCast;
 
 const MEMORY_CAPACITY: usize = 1024; // in Bytes.
 
-/// Stores a value represented as a vector of bytes, along with its type
-/// annotation. It's important to keep track of a value's type while it is
-/// being evaluated, so that we know how to interpret the bytes.
 #[derive(Debug, Clone)]
-pub struct TypedVal<'tcx> {
-    pub ty: Ty<'tcx>,
-    pub val: Vec<u8>,
+pub enum Value {
+    Int(u128),
```

Comment:
Don't we need to represent the different integer types separately? For example, imagine that I have a u32 and do `<<` on it: that could give me a different value than if it was a u64. [Similarly for overflow etc.]

Reply:
No because we always have the type information available. So when we perform arithmetic operations we downcast the operands, which are represented as `TyVal` structs, to their native rust values (by matching on the `ty` field).

Reply:
Aha, I see.

Reply:
You could argue that storing separate integer types as different enum variants would mean a less expensive lookup is needed (TypeVariants matches can be quite arduous). But as we already store the type info as part of the struct, I don't think the performance improvement would be worth the duplication of knowledge.

Reply:
OK. I guess this is something where, if we had a bytecode conversion, it would be worth doing the more fiddly case and making separate types. But you're probably right that in the current set-up it's not going to be a killer difference.

Reply:
Yeah I think that's true.